### ISSUE TITLE
AdamW implementation does not truly decouple learning rate and weight decay

### ISSUE THREAD

**Issue by leenachennuru** [NVIDIA/apex#1849]
**Describe the bug**

AdamW implementation (see [here](https://github.com/NVIDIA/apex/blob/a7de60e57f0534266841e1733262601ad76aaa74/csrc/multi_tensor_adam.cu#L333)) does not truly decouple the weight decay and learning rate parameters in line with the [adamw paper](https://arxiv.org/pdf/1711.05101). This coupling often complicates HP tuning as tuning the learning rate also changes the effective WD used to train the model.

The implementation computes the updates as

 $w_{t} = (1- \eta_{effective} \lambda) w_{t-1} - \eta_{effective} {\hat{m}_t} / {\sqrt{\hat{v}_t} + \epsilon}$ 

where $\eta_{effective} = \eta_t \eta_{max}$ with $\eta_t$ denoting the scheduler and $\eta_{max}$ the max/base LR. 

This clearly couples LR and WD and is not in line with the paper which proposes to compute the updates as 

$w_{t} = (1- \eta_t \lambda) w_{t-1} - \eta_t \eta_{max} {\hat{m}_t} / {\sqrt{\hat{v}_t} + \epsilon}$

For easier and more intuitive tuning, it would be useful to enable the completely decoupled version of AdamW via the simple fix: $\lambda = (\eta_{effective} / \eta_{max}) \lambda$ with updates: $w_{t} = (1-  \lambda) w_{t-1} - \eta_{effective} {\hat{m}_t}/{\sqrt{\hat{v}_t} + \epsilon}$.

Note: This bug also exists in implementations of [AdamW in Pytorch](https://github.com/pytorch/pytorch/blob/d921891f5788b37ea92eceddf7417d11e44290e6/torch/optim/_functional.py#L125) and [Optax](https://github.com/google-deepmind/optax/blob/2cdb89cc4935d8dc5c8a06344e7d50dc7a7419b0/optax/_src/alias.py#L640) and has already been highlighted a few times across different papers, libraries, and blogs. More links below for reference. 

1. [Mosaic ML Library](https://docs.mosaicml.com/projects/composer/en/latest/method_cards/decoupled_weight_decay.html#:~:text=The%20informed%20reader,this%20line.) 
2. [Optimi](https://optimi.benjaminwarner.dev/fully_decoupled_weight_decay/)
3. Paper: [How to set AdamW's weight decay as you scale model and dataset size](https://arxiv.org/pdf/2405.13698v1)
4. [Fabian Schaipp's blog](https://fabian-sp.github.io/posts/2024/02/decoupling/)

**Comment by timmoon10** [comment]
For better or for worse, I think "AdamW" now refers to the LR-coupled version. In addition to [PyTorch](https://github.com/pytorch/pytorch/blob/d921891f5788b37ea92eceddf7417d11e44290e6/torch/optim/_functional.py#L125) and [JAX](https://github.com/google-deepmind/optax/blob/2cdb89cc4935d8dc5c8a06344e7d50dc7a7419b0/optax/_src/alias.py#L640), I see this formulation in [Keras](https://github.com/keras-team/keras/blob/8f5592bcb61ff48c96560c8923e482db1076b54a/keras/src/optimizers/base_optimizer.py#L828C44-L829C1) (and therefore [TensorFlow](https://www.tensorflow.org/api_docs/python/tf/keras/optimizers/AdamW)), [PaddlePaddle](https://github.com/PaddlePaddle/Paddle/blob/2aebcd88c95ac8c2d917f6485c295f8837f71131/python/paddle/optimizer/adamw.py#L66), and [MXNet](https://github.com/apache/mxnet/blob/b84609d3fc73d20929c114eab95faaa56e6c5ede/python/mxnet/optimizer/adamW.py#L44). If we implement a LR-decoupled variant, we should give it a new name or make it an opt-in option.

There has been a lot of discussion in other frameworks:
- PyTorch: https://github.com/pytorch/pytorch/pull/21250#issuecomment-520289064, https://github.com/pytorch/pytorch/issues/38853, https://github.com/pytorch/pytorch/pull/3740#issuecomment-460077904
- JAX: https://github.com/google-deepmind/optax/issues/292
- Flux.jl: https://github.com/FluxML/Flux.jl/issues/2433#issuecomment-2093707811

It seems [PyTorch deliberately made the decision to use the LR-coupled variant](https://github.com/pytorch/pytorch/pull/3740#issuecomment-460077904), and that's percolated to the entire ecosystem.

**Comment by leenachennuru** [comment]
Allowing the user to invoke the fully decoupled version via either option (opt-in or another name) would be helpful. Couple more references on the potential utility of independent WD below. 

1. [Small-scale proxies for large-scale Transformer training instabilities](https://openreview.net/pdf?id=d8w0pmvXbZ)
2. [A Large-Scale Exploration of µ-Transfer](https://arxiv.org/pdf/2404.05728)
3. [u-μP: The Unit-Scaled Maximal Update Parametrization](https://arxiv.org/abs/2407.17465)
